lib/xenops_server.ml

```diff
@@ -1233,6 +1249,28 @@ let rec perform_atomic ~progress_callback ?subtask ?result (op: atomic) (t: Xeno
         List.iter (fun vusb -> VUSB_DB.remove vusb.Vusb.id) (VUSB_DB.vusbs id);
         VM_DB.remove id
     end
+  | VM_rename (id1,id2) ->
+    debug "VM.rename %s -> %s" id1 id2;
+    (* TODO: Check that there are no items in the queue for id2 *)
+    let vbds = VBD_DB.vbds id1 in
+    let vifs = VIF_DB.vifs id1 in
+    let pcis = PCI_DB.pcis id1 in
+    let vgpus = VGPU_DB.vgpus id1 in
+    let vusbs = VUSB_DB.vusbs id1 in
+    B.VM.rename id1 id2;
+    let new_key (_,y) = (id2,y) in
+    let fixup items rm add get_id set_id =
+      List.iter (fun i -> let id = get_id i in rm id; let id' = new_key id in add id' (set_id id' i)) items
+    in
+    let vm = VM_DB.read_exn id1 in
+    VM_DB.remove id1;
+    VM_DB.add id2 ({vm with Vm.id=id2});
+    fixup vbds VBD_DB.remove VBD_DB.add (fun vbd -> vbd.Vbd.id) (fun id' vbd -> {vbd with Vbd.id=id'});
+    fixup vifs VIF_DB.remove VIF_DB.add (fun vif -> vif.Vif.id) (fun id' vif -> {vif with Vif.id=id'});
+    fixup pcis PCI_DB.remove PCI_DB.add (fun pci -> pci.Pci.id) (fun id' pci -> {pci with Pci.id=id'});
+    fixup vgpus VGPU_DB.remove VGPU_DB.add (fun vgpu -> vgpu.Vgpu.id) (fun id' vgpu -> {vgpu with Vgpu.id=id'});
+    fixup vusbs VUSB_DB.remove VUSB_DB.add (fun vusb -> vusb.Vusb.id) (fun id' vusb -> {vusb with Vusb.id=id'});
```

Comment:
I would have felt better if there was a way to put this whole block in a transaction

Reply:
We might be able to use @robhoes's new atomic update stuff perhaps.

Reply:
Hmm, I don't think it's possible actually. We're already on the worker thread so no new actions affecting the domain should be running concurrently, so it's not unprotected.

Reply:
However, ops like `VIF.add` don't go into the queues...

Reply:
That would already screw up a number of other things if it were to happen at the wrong time.